extensions/common/vault/vault-hashicorp/src/test/java/org/eclipse/edc/vault/hashicorp/HashicorpVaultClientTest.java

```diff
@@ -20,217 +21,392 @@
+                assertThat(token.getTimeToLive()).isEqualTo(100);
+                verify(monitor).warning("foo-bar");
+                verify(monitor).warning("hello-world");
+            }
+
+            @Test
+            void returnsFailedResult_whenResponseIsNot200() throws IOException {
+                var response = mock(Response.class);
+                when(httpClient.execute(any(Request.class))).thenReturn(response);
+                when(response.code()).thenReturn(403);
+                when(response.isSuccessful()).thenReturn(false);
+
+                var tokenLookUpResult = vaultClient.renewToken();
+
+                assertThat(tokenLookUpResult.failed()).isTrue();
+                assertThat(tokenLookUpResult.getFailureMessages()).isEqualTo(List.of("403"));
+            }
+
+            @Test
+            void returnsFailedResult_whenIoExceptionIsThrown() throws IOException {
+                when(httpClient.execute(any(Request.class))).thenThrow(new IOException("foo-bar"));
+
+                var tokenLookUpResult = vaultClient.renewToken();
+
+                assertThat(tokenLookUpResult.failed()).isTrue();
+                assertThat(tokenLookUpResult.getFailureMessages()).isEqualTo(List.of("foo-bar"));
+            }
+
+            @Nested
+            class Scheduled {
+
+                @BeforeEach
+                void beforeEach() {
+                    reset(scheduledExecutorService);
+                }
+
+                @Test
+                void schedulesNextTokenRenewal_whenPreviousTokenRenewalWasSuccessful() throws IOException {
+                    var vaultClientSpy = spy(vaultClient);
+                    var response = mock(Response.class);
+                    var body = mock(ResponseBody.class);
+                    when(httpClient.execute(any(Request.class))).thenReturn(response);
+                    when(response.code()).thenReturn(200);
+                    when(response.isSuccessful()).thenReturn(true);
+                    when(response.body()).thenReturn(body);
+                    when(body.string())
+                            .thenReturn(
+                                    "{" +
+                                            "\"warnings\": [\"foo-bar\", \"hello-world\"]," +
+                                            "\"auth\":" +
+                                            "{" +
+                                            "\"lease_duration\": 100" +
+                                            "}" +
+                                            "}");
+                    var captor = ArgumentCaptor.forClass(Runnable.class);
+                    // break second token renewal call
+                    doNothing().when(vaultClientSpy).scheduleNextTokenRenewal(100L);
+
+                    vaultClientSpy.scheduleNextTokenRenewal(500L);
+                    verify(scheduledExecutorService).schedule(captor.capture(), eq(500L), eq(TimeUnit.SECONDS));
```

Comment:
what's going on here? please use Awaitility and a small timeout for testing?

Reply:
I'm not sure if Awaitility is the correct choice for this particular unit test

With Awaitility I would poll for a state. What is the state I can poll here? I want to verify invocations on my mocks.

Furthermore this will introduce real threads which in turn will make this an integration test since I have to test the scheduled executor service in combination with my `runnable`. With this I will also have to think about race conditions between tests etc.

While I agree with you that captors usually have different use cases I would still like to go with this solutions as it's the most elegant solution

Other options could be:
- Extract the `runnable` in a separate package-private method or nested class and test it separately. But this will go against the guidelines since it's a package-private code for testing only
- Extract the `runnable` in a separate public class. This doesn't feel right as well

Reply:
I would getting more familiar with Awaitility. It has the `untilAsserted()` method that takes a lambda, in which you can verify a mock.

To make it clear: ArgumentCaptors have a very different use case, and you **will not** abuse them in this way. 

One way you could do this is as follows:
- the `HashicorpVaultClient` takes the executor and the schedule period in its CTor
- in the test you pass a very small timeout, eg. 5 ms
- with awaitility you wait until the mock verifies, an assertion is true, etc.